src/network/peers.rs

```diff
@@ -671,7 +671,13 @@ impl<N: Network, E: Environment> Peers<N, E> {
     ///
     /// Sends the given message to every connected peer, excluding the sender.
     ///
-    async fn propagate(&self, sender: SocketAddr, message: Message<N, E>) {
+    async fn propagate(&self, sender: SocketAddr, mut message: Message<N, E>) {
+        // Perform ahead-of-time, non-blocking serialization just once for applicable objects.
+        if let Message::UnconfirmedBlock(_, _, ref mut data) = message {
+            let serialized_block = Data::serialize(data.clone()).await.expect("Block serialization is bugged");
+            let _ = std::mem::replace(data, Data::Buffer(serialized_block));
+        }
+
```

Comment:
note: I'm not altering the other point where ahead-of-time block serialization is currently performed in order to limit the scope of changes; the only downside of doing it that way is an extra `clone()` per block when propagating.

Reply:
We could probably `mem::take` but this will require some default impl for the block, which we don't currently have.